Add Windows Node 14 Testing to CircleCI

bendvc opened the pull request:
## Problem

After moving to the Salesforce Commerce Cloud organization we lost our windows testing solution (AppVeyor). Fortunately CircleCi now supports windows testing. The goal of this PR is to update the CircleCI configuration to enable windows testing.

## Solution

Updated the `.circleci/config.yml` to include `testNode12Windows` and `setup_windows` jobs and commands respectively.

## Known Issues

Ideally we support both Node 14 and Node 12 as those are the versions the SDK supports. Unfortunately the windows image we use to do our testing is having issues using the version of NVM installed. In short after using nvm to change the version of node, the `npm` and `node` commands do not work. What makes it harder to debug is that this doesn't happen when you ssh into the VM, and furthermore it seems like the environment variables are the same. So this PR doesn't support node 12, and we'll try implementing that at a later time.

A [GUS ticket](https://gus.lightning.force.com/lightning/_classic/%2Fa07EE00000B0m4mYAB) was created to follow up for Node 12 support at a later time.

bendvc committed:
Add basic windows circle ci job

bendvc committed:
Remove "build" key

bendvc committed:
Add checkout and install steps

bendvc committed:
No capitals in command names

bendvc committed:
Add testing step, and caching or npm to make it faster?

bendvc committed:
fix syntax error

bendvc committed:
Whoops, mixed up my npm and node version numbers

bendvc committed:
Use `npm i` instead.

bendvc committed:
Disable setup flow, to make manual testing faster

bendvc committed:
Reduce command set to bare bones

bendvc committed:
Attempt to fix issue with switching node versions

bendvc committed:
nvm alias doesn't work

bendvc committed:
print env var before and after using nvm

bendvc committed:
remove old node before using nvm

bendvc committed:
Fix up removal of symlink

bendvc committed:
Use older version of window orb to see if that fixes nvm issue

bendvc committed:
Trying a newer version

bendvc committed:
try a difference shell

bendvc committed:
Lets try bash

bendvc committed:
Set the home path after changing node versions

bendvc committed:
Update config.yml

bendvc committed:
simplify

bendvc committed:
Update config.yml

bendvc committed:
npm install works, does npm ci

bendvc committed:
Ensure we build the project

bendvc committed:
User large resource machine

I suspect that there is some memory thrashing going on during translation compilation, lets try a better spec machine.

bendvc committed:
Remove build step.

bendvc committed:
Merge branch 'develop' into circle-ci-windows

bendvc committed:
Merge branch 'develop' into circle-ci-windows

bendvc committed:
Added script sanity check and clean check after running build and tests for windows.

bendvc committed:
Remove sanity check on scripts

They currently don't work on windows.

bendvc committed:
Setup multiple windows orbs

testing their nodes versions

bendvc committed:
Default in windows ord is actually node 14 :)

adamraya commented on .circleci/config.yml:
```diff
@@ -150,6 +168,16 @@ jobs:
+      - checkout
+      - setup_windows
+      - runtests:
+          enableIntegrationTests: false
```
Before we were running the integration tests on the master branches on Windows:
https://github.com/mobify/mobify-platform-sdks/blob/salesforce-develop/appveyor.yml#L59-L62

Now we're not running any integration test on Windows. Is that intentional?

bendvc replied:
Good catch, I must has copied the `enableIntegrationTests: false` parameter from another job. I'll remove that so that the integration tests are run on those branches since `enableIntegrationTests` defaults to `true`.

adamraya reviewed (commented):
I see an error running the `test:max-file-size` script on Windows, but the job still succeeds:
https://app.circleci.com/pipelines/github/SalesforceCommerceCloud/pwa-kit/903/workflows/ee85c72e-a284-4398-b98c-71230e410053/jobs/2679?invite=true#step-105-31

Probably related to the path `build/main.js` not being cross-platform. 

```
scaffold-pwa: > scaffold-pwa@1.1.0-dev test:max-file-size C:\Users\circleci\project\packages\pwa
scaffold-pwa: > bundlesize
scaffold-pwa: There is no matching file for build/main.js in C:\Users\circleci\project\packages\pwa
scaffold-pwa: npm ERR! code ELIFECYCLE
scaffold-pwa: npm ERR! errno 1
scaffold-pwa: npm ERR! scaffold-pwa@1.1.0-dev test:max-file-size: `bundlesize`
scaffold-pwa: npm ERR! Exit status 1
scaffold-pwa: npm ERR!
scaffold-pwa: npm ERR! Failed at the scaffold-pwa@1.1.0-dev test:max-file-size script.
scaffold-pwa: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
scaffold-pwa: npm ERR! A complete log of this run can be found in:
scaffold-pwa: npm ERR!     C:\Users\circleci\AppData\Roaming\npm-cache\_logs\2021-09-27T18_00_42_979Z-debug.log
lerna ERR! npm run test:max-file-size exited 1 in 'scaffold-pwa'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! mobify-platform-sdks@1.1.0-dev test:max-file-size: `lerna run --stream test:max-file-size`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the mobify-platform-sdks@1.1.0-dev test:max-file-size script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\circleci\AppData\Roaming\npm-cache\_logs\2021-09-27T18_00_43_020Z-debug.log
```

adamraya reviewed (commented):
This test FAILS on Windows (just a timeout, not Windows related probably), but the "Run Tests" job still exits with a zero exit. i.e. The `testNode14Windows` job succeeds with tests failing: 
https://app.circleci.com/pipelines/github/SalesforceCommerceCloud/pwa-kit/903/workflows/ee85c72e-a284-4398-b98c-71230e410053/jobs/2679?invite=true#step-105-1650

adamraya reviewed (commented):
Do we know if the e2e test works on Windows?
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/9d6a276b3294b2deb2b568df508d14a2bd8082c0/.circleci/config.yml#L109-L112
Once we merge this into `develop`, the nightly build will run the e2e test on Windows:
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/9d6a276b3294b2deb2b568df508d14a2bd8082c0/.circleci/config.yml#L311-L318

I'm not sure if it's going to be necessary here but in previous builds, we needed to upgrade the Chrome version:
https://github.com/mobify/mobify-platform-sdks/blob/salesforce-develop/appveyor.yml#L25-L30

I tried to SSH into the build but for some reason, after the session starts the terminal doesn't show anything. Maybe we need to include some instructions in the readme on how to SSH into the Windows builds.

bendvc committed:
Merge branch 'develop' into circle-ci-windows

bendvc committed:
Update config.yml

- Remove bundle size check as it's not cross platform compatible.
- Ensure integration tests run.

bendvc commented:
> I see an error running the `test:max-file-size` script on Windows, but the job still succeeds: https://app.circleci.com/pipelines/github/SalesforceCommerceCloud/pwa-kit/903/workflows/ee85c72e-a284-4398-b98c-71230e410053/jobs/2679?invite=true#step-105-31
> 
> Probably related to the path `build/main.js` not being cross-platform.
> 
> ```
> scaffold-pwa: > scaffold-pwa@1.1.0-dev test:max-file-size C:\Users\circleci\project\packages\pwa
> scaffold-pwa: > bundlesize
> scaffold-pwa: There is no matching file for build/main.js in C:\Users\circleci\project\packages\pwa
> scaffold-pwa: npm ERR! code ELIFECYCLE
> scaffold-pwa: npm ERR! errno 1
> scaffold-pwa: npm ERR! scaffold-pwa@1.1.0-dev test:max-file-size: `bundlesize`
> scaffold-pwa: npm ERR! Exit status 1
> scaffold-pwa: npm ERR!
> scaffold-pwa: npm ERR! Failed at the scaffold-pwa@1.1.0-dev test:max-file-size script.
> scaffold-pwa: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> scaffold-pwa: npm ERR! A complete log of this run can be found in:
> scaffold-pwa: npm ERR!     C:\Users\circleci\AppData\Roaming\npm-cache\_logs\2021-09-27T18_00_42_979Z-debug.log
> lerna ERR! npm run test:max-file-size exited 1 in 'scaffold-pwa'
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! mobify-platform-sdks@1.1.0-dev test:max-file-size: `lerna run --stream test:max-file-size`
> npm ERR! Exit status 1
> npm ERR!
> npm ERR! Failed at the mobify-platform-sdks@1.1.0-dev test:max-file-size script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\circleci\AppData\Roaming\npm-cache\_logs\2021-09-27T18_00_43_020Z-debug.log
> ```

I've removed the bundle-size test. The original appveyor didn't rung that and we can rely on the other non-windows builds to inform us of failures.

bendvc commented:
> Do we know if the e2e test works on Windows?
> 
> https://github.com/SalesforceCommerceCloud/pwa-kit/blob/9d6a276b3294b2deb2b568df508d14a2bd8082c0/.circleci/config.yml#L109-L112
> 
> 
> Once we merge this into `develop`, the nightly build will run the e2e test on Windows:
> https://github.com/SalesforceCommerceCloud/pwa-kit/blob/9d6a276b3294b2deb2b568df508d14a2bd8082c0/.circleci/config.yml#L311-L318
> 
> I'm not sure if it's going to be necessary here but in previous builds, we needed to upgrade the Chrome version: https://github.com/mobify/mobify-platform-sdks/blob/salesforce-develop/appveyor.yml#L25-L30
> 
> I tried to SSH into the build but for some reason, after the session starts the terminal doesn't show anything. Maybe we need to include some instructions in the readme on how to SSH into the Windows builds.

Ok, so to answer these questions. 

1. The tests do run when invoked manually, so we can rely on circleci's the conditional working as it works for other jobs. See the output of the command being run ssh'd into the windows box below. 

2. I believe that chrome is installed for lighthouse reasons, not for e2e as you can see on [this](https://app.circleci.com/pipelines/github/SalesforceCommerceCloud/pwa-kit/977/workflows/63d1a2d7-de9b-4865-8fd5-318630405380/jobs/2913) build, (search " Executing command in 1 package: "npm run test:e2e-ci" ") to see that even the non-windows builds are using the headless electron browser for e2e. 

3. That is a known bug with CircleCi that if you ssh using bash and your chosen shell it will hang, you have to use powershell or cmd.


```
lerna notice cli v3.22.1
lerna info ci enabled
lerna info Executing command in 1 package: "npm run test:e2e-ci"
scaffold-pwa: > scaffold-pwa@1.2.0-dev test:e2e-ci C:\Users\circleci\project\packages\pwa
scaffold-pwa: > node ./scripts/run-e2e.js
scaffold-pwa: [Poll] Waiting for Express app to start (Attempt 1)
scaffold-pwa: [Poll] No response. 204966ms remaining
scaffold-pwa: [Poll] Waiting for Express app to start (Attempt 2)
scaffold-pwa: [Poll] No response. 194949ms remaining
scaffold-pwa: [Poll] Waiting for Express app to start (Attempt 3)
scaffold-pwa: [Poll] No response. 184927ms remaining
scaffold-pwa: [Poll] Waiting for Express app to start (Attempt 4)
scaffold-pwa: [Poll] No response. 174912ms remaining
scaffold-pwa: [Poll] Waiting for Express app to start (Attempt 5)
scaffold-pwa: [Poll] Server is running
scaffold-pwa: > scaffold-pwa@1.2.0-dev test:e2e C:\Users\circleci\project\packages\pwa
scaffold-pwa: > cypress run
scaffold-pwa: It looks like this is your first time using Cypress: 6.9.1
scaffold-pwa: [19:43:33]  Verifying Cypress can run C:\Users\circleci\AppData\Local\Cypress\Cache\6.9.1\Cypress [started]
scaffold-pwa: [19:43:37]  Verifying Cypress can run C:\Users\circleci\AppData\Local\Cypress\Cache\6.9.1\Cypress [completed]
scaffold-pwa: 
scaffold-pwa: Opening Cypress...
scaffold-pwa: Couldn't find tsconfig.json. tsconfig-paths will be skipped
scaffold-pwa: ====================================================================================================
scaffold-pwa:   (Run Starting)
scaffold-pwa:   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
scaffold-pwa:   │ Cypress:    6.9.1                                                                              │
scaffold-pwa:   │ Browser:    Electron 87 (headless)                                                             │
scaffold-pwa:   │ Specs:      1 found (pages\home.spec.js)                                                       │
scaffold-pwa:   └────────────────────────────────────────────────────────────────────────────────────────────────┘
scaffold-pwa: ────────────────────────────────────────────────────────────────────────────────────────────────────
scaffold-pwa:                                                                                                     
scaffold-pwa:   Running:  pages\home.spec.js                                                              (1 of 1)
scaffold-pwa: Browserslist: caniuse-lite is outdated. Please run:
scaffold-pwa: npx browserslist@latest --update-db
scaffold-pwa: Why you should do it regularly:
scaffold-pwa: https://github.com/browserslist/browserslist#browsers-data-updating
scaffold-pwa: 
scaffold-pwa:   Home Page
scaffold-pwa:     √ loads successfully (2850ms)
scaffold-pwa:   1 passing (3s)
scaffold-pwa:   (Results)
scaffold-pwa:   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
scaffold-pwa:   │ Tests:        1                                                                                │
scaffold-pwa:   │ Passing:      1                                                                                │
scaffold-pwa:   │ Failing:      0                                                                                │
scaffold-pwa:   │ Pending:      0                                                                                │
scaffold-pwa:   │ Skipped:      0                                                                                │
scaffold-pwa:   │ Screenshots:  0                                                                                │
scaffold-pwa:   │ Video:        true                                                                             │
scaffold-pwa:   │ Duration:     2 seconds                                                                        │
scaffold-pwa:   │ Spec Ran:     pages\home.spec.js                                                               │
scaffold-pwa:   └────────────────────────────────────────────────────────────────────────────────────────────────┘
scaffold-pwa:   (Video)
scaffold-pwa:   -  Started processing:  Compressing to 32 CRF                                                     
scaffold-pwa:   -  Finished processing: C:\Users\circleci\project\packages\pwa\cypress\videos\pages    (0 seconds)
scaffold-pwa:                           \home.spec.js.mp4                                                         
scaffold-pwa: ====================================================================================================
scaffold-pwa:   (Run Finished)
scaffold-pwa:        Spec                                              Tests  Passing  Failing  Pending  Skipped  
scaffold-pwa:   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
scaffold-pwa:   │ ✔  pages\home.spec.js                       00:02        1        1        -        -        - │
scaffold-pwa:   └────────────────────────────────────────────────────────────────────────────────────────────────┘
scaffold-pwa:     ✔  All specs passed!                        00:02        1        1        -        -        -  
scaffold-pwa: 
lerna success run Ran npm script 'test:e2e-ci' in 1 package in 72.7s:
lerna success - scaffold-pwa


```

bendvc committed:
testing

bendvc commented:
> This test FAILS on Windows (just a timeout, not Windows related probably), but the "Run Tests" job still exits with a non-zero exit. i.e. The `testNode14Windows` job succeeds with tests failing: https://app.circleci.com/pipelines/github/SalesforceCommerceCloud/pwa-kit/903/workflows/ee85c72e-a284-4398-b98c-71230e410053/jobs/2679?invite=true#step-105-1650

I'm going to investigate this. It's either an issue with circleci or how we are running our commands. I'll keep you updated.

## Update

It looks like the default pipefail settings aren't being applied on windows machines (-eo pipefail). This means that even though commands fail, the subsequent commands are running. I've tested that this is the case but inserting a failing command (mkdir .) in the step with and without explicitly setting `-eo pipefail` via "set -eo pipefail). Hopefully it all works out now. You can read about this setting more [here](https://circleci.com/docs/2.0/configuration-reference/#default-shell-options).

bendvc committed:
another test for status code

bendvc committed:
more testing

bendvc committed:
more testing

checking what the pipefail settings are

bendvc committed:
Double check that set -eo has to be set for windows.

bendvc committed:
Fix pipefile set command

bendvc committed:
Remove failing command test

adamraya reviewed (approved):
LGTM
